core/matrix/batch_dense.cpp

```diff
@@ -124,11 +116,72 @@ Dense<ValueType>::Dense(std::shared_ptr<const Executor> exec,
 {}
 
 
+template <typename ValueType>
+Dense<ValueType>* Dense<ValueType>::apply(
+    ptr_param<const MultiVector<ValueType>> b,
+    ptr_param<MultiVector<ValueType>> x)
+{
+    this->validate_application_parameters(b.get(), x.get());
```

Comment:
It is not necessary to implement both the const and non-const apply. The non-const can just call the const version and return `this`. Same for the advanced apply